File: python-examples/07-lifecycle-hooks/agents.py
"""Agents with lifecycle hooks attached."""

from polos import Agent, max_steps, MaxStepsConfig

from tools import search, calculate
from hooks import (
    log_start,
    log_end,
    log_step_start,
    log_step_end,
    log_tool_start,
    log_tool_end,
    validate_input,
)


# Agent with full lifecycle logging
logged_agent = Agent(
    id="logged_agent",
    provider="openai",
    model="gpt-4o-mini",
    system_prompt="""You are a helpful assistant with access to search and calculator tools.
Use these tools to help answer user questions.""",
    tools=[search, calculate],
    # Lifecycle hooks
    on_start=[validate_input, log_start],  # Multiple hooks run in order
    on_end=[log_end],
    on_agent_step_start=[log_step_start],
    on_agent_step_end=[log_step_end],
    on_tool_start=[log_tool_start],
    on_tool_end=[log_tool_end],
    stop_conditions=[
        max_steps(MaxStepsConfig(limit=5)),
    ],
)


# Agent with just start/end logging
simple_logged_agent = Agent(
    id="simple_logged_agent",
    provider="openai",
    model="gpt-4o-mini",
    system_prompt="""You are a helpful assistant.""",
    on_start=[log_start],
    on_end=[log_end],
    stop_conditions=[
        max_steps(MaxStepsConfig(limit=5)),
    ],
)


# Agent with input validation
validated_agent = Agent(
    id="validated_agent",
    provider="openai",
    model="gpt-4o-mini",
    system_prompt="""You are a helpful assistant.""",
    on_start=[validate_input],  # Will reject empty or overly long prompts
    stop_conditions=[
        max_steps(MaxStepsConfig(limit=5)),
    ],
)

File: typescript-examples/07-lifecycle-hooks/agents.ts
/**
 * Agents with lifecycle hooks attached.
 */

import { defineAgent, maxSteps } from '@polos/sdk';
import { openai } from '@ai-sdk/openai';
import { search, calculate } from './tools.js';
import {
  logStart,
  logEnd,
  logStepStart,
  logStepEnd,
  logToolStart,
  logToolEnd,
  validateInput,
} from './hooks.js';

// Agent with full lifecycle logging
export const loggedAgent = defineAgent({
  id: 'logged_agent',
  model: openai('gpt-4o-mini'),
  systemPrompt:
    'You are a helpful assistant with access to search and calculator tools. ' +
    'Use these tools to help answer user questions.',
  tools: [search, calculate],
  // Lifecycle hooks
  onStart: [validateInput, logStart], // Multiple hooks run in order
  onEnd: [logEnd],
  onAgentStepStart: [logStepStart],
  onAgentStepEnd: [logStepEnd],
  onToolStart: [logToolStart],
  onToolEnd: [logToolEnd],
  stopConditions: [maxSteps({ count: 5 })],
});

// Agent with just start/end logging
export const simpleLoggedAgent = defineAgent({
  id: 'simple_logged_agent',
  model: openai('gpt-4o-mini'),
  systemPrompt: 'You are a helpful assistant.',
  onStart: [logStart],
  onEnd: [logEnd],
  stopConditions: [maxSteps({ count: 5 })],
});

// Agent with input validation
export const validatedAgent = defineAgent({
  id: 'validated_agent',
  model: openai('gpt-4o-mini'),
  systemPrompt: 'You are a helpful assistant.',
  onStart: [validateInput], // Will reject empty or overly long prompts
  stopConditions: [maxSteps({ count: 5 })],
});
